### ISSUE TITLE
Export for sample tracks submitted through the eBMS app

### ISSUE THREAD

**Issue by DavidRoy** [BiologicalRecordsCentre/ABLE#151]
The eBMS captures a route associated with samples (see example below).  Requirement is for a user to be able to download the tracks.

One options is to extend the existing Transect route download report at bottom of https://butterfly-monitoring.net/my-sites:
![dl](https://user-images.githubusercontent.com/2699478/66993325-46e1c180-f0c3-11e9-9fe7-4da76bcd377b.png)

![track](https://user-images.githubusercontent.com/2699478/66992922-9e336200-f0c2-11e9-8e53-42d44d0d653b.png)

**Comment by burkmarr** [comment]
@DavidRoy - I'm not sure that I understand what is required here beyond what is already available. I just tried the ```SHP format GIS section download``` and I got the sections geometry.

I did this for the user we recently discussed via email and got the same results as I extracted - using the more long-winded method - direct from the warehouse.

Am I misunderstanding something?

**Comment by DavidRoy** [comment]
@burkmarr yes, the current functionality work.  What is required here is an equivalent report for the mobile app GIS data - survey.id=565

The current reports are for eBMS transects, survey.id = 562

**Comment by Gary-van-Breda** [comment]
Just to clarify what the existing report does: There is no dependance on the survey.
It provides the geometries of the child locations of the sites (of a configurable location type, currently Transects), that are allocated to the person via the "CMS User ID" attribute (again configurable, is the "Country" attribute via the Scheme for scheme coordinators)   
What location type is being used for the App sites?

**Comment by DavidRoy** [comment]
I’m not sure what location type is set for app samples - likely the warehouse default?

**Comment by Gary-van-Breda** [comment]
I can llok up - do you have the name of one of them?

**Comment by DavidRoy** [comment]
They don’t have names but will be under survey.id = 575

**Comment by Gary-van-Breda** [comment]
OK - I didn't under quite twig that the path is stored in the geometry on the sample record, and they do not use location records.  
This will need a new report.

**Comment by DavidRoy** [comment]
These paths are not necessarily repeated so are different from transect routes which are fixed and regularly recorded.
@Gary-van-Breda is there anything 'bad' about this approach?

**Comment by Gary-van-Breda** [comment]
Not at all: it's just different.

**Comment by burkmarr** [comment]
I've tried an approach based on the Transect Sites report referenced above (https://butterfly-monitoring.net/my-sites) because it seemed like it would be the simplest place to start. However, I have been unable to get the GIS download link to work because I cannot find a way to pass the Indicia User ID to a download report via a Get request.

The page transects sites page doesn't need to do this - instead it passes the CMS user ID which is accessible in a Get request through the {user_id} token. For example the KML download link at the bottom of that page looks like this:
```
<a href="{warehouseRoot}index.php/services/report/requestReport?
report=reports_for_prebuilt_forms/UKBMS/ukbms_section_gis_download.xml
&reportSource=local&mode=kml&{nonce}&{auth}
&location_type_id=777&locattrs=234&attr_location_234={user_id}">
KML format GIS Section Download</a>
```
It can do this because the UKBMS appears to store the CMS user ID as a custom attribute on locations it makes. But there's no equivalent way to identify samples (which is where the timed counts app stores the relevant geometry) using a CMS user ID that I can see.

I thought I would be able to specify the Indicia user ID on the Get request string using the token {indicia_user_id} because I know that token operates as a parameter within an iForm reports page (just like {user_id} does). However, it doesn't get translated to the Indicia user ID, as you can see if you try the link a the bottom of the report page here: https://butterfly-monitoring.net/app-sites-test

Does anyone know (@Gary-van-Breda @johnvanbreda) how I can identify the Indicia user via the Get request?

Another approach, and perhaps a better one, might be to enhance the current 'My app timed counts' summary page (https://butterfly-monitoring.net/elastic/my-records) since there is a tab on there already which summarises a user's samples (https://butterfly-monitoring.net/elastic/my-records). The page also has a Download tab but, at the moment, this only offers a download of the occurrences. The page queries the Elasticsearch Index rather than the Warehouse DB and I don't know if it is possible to retrieve sample geometry from it. Is it possible @johnvanbreda?

**Comment by DavidRoy** [comment]
thanks Rich.  For performance reasons, I agree that it would be benefitial to export via the ElasticSearch Index.  Also, i agree that it makes sense to add to the my-records page.

**Comment by Gary-van-Breda** [comment]
The report grid footer translations are slightly different to the normal ones: it was done quite a while ago. They can be found in the report_grid function in the report_helper.php: the indicia_user_id can be accessed using the '{iUserID}' token.

**Comment by burkmarr** [comment]
Thanks @Gary-van-Breda - that's what I needed to know. 

@DavidRoy - I've completed the approach I took yesterday, so now a user can download the GIS tracks of their timed counts from this page: https://butterfly-monitoring.net/my-timed-counts-samples-gis-download.

So if users have been asking you for this, you can point them to this page. There's currently no link to it from the main menu. Let me know if you want to add that in. I've made a pull request to add the reports to those stored within the warehouse repository: https://github.com/Indicia-Team/warehouse/pull/342. (But they are already deployed so we are not waiting for this to be merged.)

Currently the tracks can be downloaded as KML or GPX. To provide a shapefile download I would have to do some more learning and possibly some work on GeoServer. So, unless this is a priority for some reason, I propose to leave it until the general way forward is clarified. If you want me to investigate the approach of extending the Elasticsearch page to add a GIS download - subject to John confirming that this is possble from our current ES index - let me know. It's likely to be quite time-consuming for me because I would need to do some learning on it and there are probably other priorities at the moment.

**Comment by Gary-van-Breda** [comment]
The SHP file downloads used elsewhere on ABLE require a view to be set up in the Postgres database on the warehouse (currently it uses the same view as UKBMS, and that is in the ukbms schema), and a layer to be set up in the Geoserver to point to the view.

**Comment by johnvanbreda** [comment]
@burkmarr @DavidRoy as long as the geometry of the path is stored in the sample the occurrence is directly attached to, then yes, this information is stored along with each occurrence in Elasticsearch (in the location.geom field). The geometries associated with parent samples are not stored at it stands but could be added, though at the cost of extra overhead of course.

**Comment by burkmarr** [comment]
Thanks @johnvanbreda and @Gary-van-Breda.

**Comment by DavidRoy** [comment]
@burkmarr bumping this issue from a while ago

**Comment by burkmarr** [comment]
@DavidRoy to tackle this with ES requires our ES API to be enhanced to deliver formats other than CSV. I am looking at this as part of the ES download work I'm tasked with so can return to this issue once I've progressed that. In the meantime, don't forget that I created the page referred to above (https://butterfly-monitoring.net/my-timed-counts-samples-gis-download) where people can download their timed count tracks now - using non-ES technology. Previously this was only KML or GPX but today I've added Shapefile to it (thanks to the Gary's comment which pointed the way to doing that). There is no link to this page from the website though so let me know if you want me to add one.
